src/utils.rs

```diff
@@ -146,54 +148,84 @@ impl Console {
+        epoll
+            .ctl(
+                ControlOperation::Add,
+                self.fd,
+                EpollEvent::new(EventSet::IN, self.fd as u64),
+            )
+            .map_err(|e| {
+                new_nitro_cli_failure!(
+                    &format!("Failed to add fd to epoll: {:?}", e),
+                    NitroCliErrorEnum::EpollError
+                )
+            })?;
+
+        // If the function call provides a disconnect timeout, create a timerfd,
+        // arm it and then add it to epoll
+        if let Some(disconnect_timeout) = disconnect_timeout_sec {
+            // Create timerfd
+            let mut timerfd = TimerFd::new().map_err(|e| {
+                new_nitro_cli_failure!(
+                    &format!("Failed to initialize timerfd: {:?}", e),
+                    NitroCliErrorEnum::EpollError
+                )
+            })?;
+
+            // Arm timerfd with disconnect_timeout seconds
+            timerfd
+                .reset(Duration::from_secs(disconnect_timeout), None)
+                .map_err(|e| {
+                    new_nitro_cli_failure!(
+                        &format!("Failed to arm timerfd: {:?}", e),
+                        NitroCliErrorEnum::EpollError
+                    )
+                })?;
+
+            // Add timerfd fd to epoll
+            let timerfd_fd = timerfd.into_raw_fd();
             epoll
-                .ctl(ControlOperation::Add, self.fd, epoll_event)
+                .ctl(
+                    ControlOperation::Add,
+                    timerfd_fd,
+                    EpollEvent::new(EventSet::IN, timerfd_fd as u64),
+                )
                 .map_err(|e| {
                     new_nitro_cli_failure!(
                         &format!("Failed to add fd to epoll: {:?}", e),
                         NitroCliErrorEnum::EpollError
                     )
                 })?;
-
-            // Disconnect timeout in microseconds
-            epoll_timeout_us = (disconnect_timeout_sec.unwrap_or(0) * 1000 * 1000) as i128;
         }
 
         loop {
-            if disconnect_timeout_sec.is_some() {
-                start_epoll_time = Instant::now();
-
-                // Use epoll_wait to exit the blocking state when the fd is ready to be read
-                // or when the disconnect time has passed
-                let num_events = epoll
-                    .wait((epoll_timeout_us / 1000) as i32, &mut events)
-                    .map_err(|e| {
-                        new_nitro_cli_failure!(
-                            &format!("Failed to wait on epoll: {:?}", e),
-                            NitroCliErrorEnum::EpollError
-                        )
-                    })?;
-
-                // If the timeout expires, no event happend and the console disconnects
-                if num_events == 0 {
+            // Wait for kernel notification that one of the fds is available
+            epoll.wait(-1, &mut events).map_err(|e| {
```

Comment:
I don't really see that this is simpler to understand, both timeout and a timer_fd provide the same functionality, but timer_fd needs to be declared and armed, while timeout abstracts this. There were some improvements to be done to how things were handled previously, for sure, but I don't understand how the timerfd makes things easier to understand this case, and the commit message doesn't really clarify it for me either.

Reply:
I see a small benefit in eliminating the logic of manually decreasing the value of timer in a loop.

Reply:
I think that letting the kernel to manage the timer and notify the process when the timer expired is better than managing the timer in user space. Handling different corner cases with the math done for keeping track of elapsed time might be difficult and produce cluttered code.

Also, this version is more scalable, if you ever want to add another fd to epoll (eg. EventFd).